src/asm_writing/rewriter.cpp

```diff
@@ -1148,6 +1146,20 @@ void Rewriter::_setupCall(bool has_side_effects, llvm::ArrayRef<RewriterVar*> ar
 #endif
 }
 
+void Rewriter::_callOptimalEncoding(assembler::Register tmp_reg, void* func_addr) {
```

Comment:
should be fine because like you wrote we are selecting the ICSlot at the start of the emitting phase.
And we were already doing this relative call optimization inside `_call` this only extended it to other places too.

As for the bjit being reentrant: the `JitCodeBlock` has a `is_currently_writing` which makes sure that we don't start another bjit emitter for the same function. So this should be fine too :+1: